doc/source/train/user-guides/checkpoints.rst

```diff
@@ -233,6 +233,57 @@ Here is an example of distributed checkpointing with PyTorch:
     rank-specific filenames already, so you usually do not need to worry about this.
 
 
+.. _train-checkpoint-upload-modes:
+
+Checkpoint Upload Modes
+-----------------------
+
+By default, when you call :func:`~ray.train.report`, Ray Train synchronously pushes
+your checkpoint from ``checkpoint.path`` on local disk to ``checkpoint_dir_name`` on
+your ``storage_path``. This is equivalent to calling :func:`~ray.train.report` with
+:class:`~ray.train.CheckpointUploadMode` set to ``ray.train.CheckpointUploadMode.SYNC``.
+
+.. literalinclude:: ../doc_code/checkpoints.py
+    :language: python
+    :start-after: __checkpoint_upload_mode_sync_start__
+    :end-before: __checkpoint_upload_mode_sync_end__
+
+Asynchronous Checkpoint Uploading
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+You may want to upload your checkpoint asynchronously instead so that
+the next training step can start in parallel. If so, you should use
+``ray.train.CheckpointUploadMode.ASYNC``. This is helpful for larger
+checkpoints that might take longer to upload, but might add unnecessary
+complexity if you want to immediately upload a small checkpoint.
+
+.. literalinclude:: ../doc_code/checkpoints.py
+    :language: python
+    :start-after: __checkpoint_upload_mode_async_start__
+    :end-before: __checkpoint_upload_mode_async_end__
+
+.. figure:: ../images/sync_vs_async_checkpointing.png
+
+    Synchronous vs asynchronous checkpoint uploading. Right now, each ``report`` blocks
+    until the previous ``report``\'s checkpoint upload completes before starting a new
+    checkpoint upload thread. We do this to avoid accumulating too many upload threads
+    and potentially running out of memory.
```

Comment:
The threading comments can be moved to be part of the regular docs, as opposed to the caption of the image.